### ISSUE TITLE
2.1.1 regression: @inaccessible + @requires fails at runtime

### ISSUE THREAD

**Issue by lennyburdette** [apollographql/federation#2152]
Reproduction: https://stackblitz.com/edit/basic-federation-ywtnvt?file=one.js,two.js (with operation `{ one { b } }`)
Version: 2.1.1 and 2.1.2 (works with 2.0.5)

With subgraphs:
```graphql
extend schema
  @link(url: "https://specs.apollo.dev/federation/v2.0",
        import: ["@key", "@shareable", "@inaccessible"])

type Query {
  one: One
}

type One @key(fields: "id") {
  id: ID!
  a: String @inaccessible 
}
```
```graphql
extend schema
  @link(url: "https://specs.apollo.dev/federation/v2.0",
        import: ["@key", "@external", "@requires"])

type One @key(fields: "id") {
  id: ID!
  a: String @external
  b: String @requires(fields: "a" )
}
```
we get a runtime error: 
```
{
  "errors": [
    {
      "message": "Cannot add selection of field \"One.a\" to selection set of parent type \"One\" (that does not declare that field)",
      "extensions": {
        "code": "INVALID_GRAPHQL",
        "exception": {
          "stacktrace": [
            "GraphQLError: Cannot add selection of field \"One.a\" to selection set of parent type \"One\" (that does not declare that field)",
            "    at Object.err (/home/projects/basic-federation-ywtnvt/node_modules/@apollo/federation-internals/dist/error.js:11:32)",
            "    at validate (/home/projects/basic-federation-ywtnvt/node_modules/@apollo/federation-internals/dist/operations.js:12:46)",
            "    at Field.updateForAddingTo (/home/projects/basic-federation-ywtnvt/node_modules/@apollo/federation-internals/dist/operations.js:115:9)",
            "    at FieldSelection.updateForAddingTo (/home/projects/basic-federation-ywtnvt/node_modules/@apollo/federation-internals/dist/operations.js:1010:41)",
            "    at SelectionSet.add (/home/projects/basic-federation-ywtnvt/node_modules/@apollo/federation-internals/dist/operations.js:707:33)",
            "    at SelectionSet.mergeIn (/home/projects/basic-federation-ywtnvt/node_modules/@apollo/federation-internals/dist/operations.js:698:18)",
            "    at addSelectionOrSelectionSet (/home/projects/basic-federation-ywtnvt/node_modules/@apollo/query-planner/dist/buildPlan.js:1326:22)",
            "    at eval (/home/projects/basic-federation-ywtnvt/node_modules/@apollo/query-planner/dist/buildPlan.js:439:21)",
            "    at selectionSetOfPath (/home/projects/basic-federation-ywtnvt/node_modules/@apollo/federation-internals/dist/operations.js:922:9)",
            "    at FetchGroup.mergeInInternal (/home/projects/basic-federation-ywtnvt/node_modules/@apollo/query-planner/dist/buildPlan.js:435:74)"
          ]
        }
      }
    }
  ]
}
```

**Comment by clenfest** [comment]
Looks like the regression was introduced in https://github.com/apollographql/federation/pull/1958, which was in between `2.1.0-alpha.2` and `2.1.0-alpha.3`.

**Comment by clenfest** [comment]
Issue appears to be here https://github.com/apollographql/federation/blob/main/internals-js/src/operations.ts#L201-L230

Previous code was 
```typescript
  if (selectionParent.name !== fieldParent.name) {
...
}
  return this;
```

@pcmanus Can we just change it back to use `.name`?

**Comment by pcmanus** [comment]
> Can we just change it back to use .name?

No, that's incorrect and that's why it was changed in the first place. One of the main goal of `updateForAddingTo` is to make sure we don't end up mistakenly creating Frankenstein `SelectionSet` whose elements points to _different_ underlying schema (which is obviously wrong, and also quite painful to debug), and that's what the change made to that method avoids. Changing this would silence the assertion in this case, but only by leaving the underlying issue the assertion is pointing to, and that would almost surely create some subtle issues.

The failing assertion points to a genuine issue within `FetchGroup.mergeInInternal` where we should be building a selection set for a subgraph, but instead create one against the supergraph API schema because we base it on the `path` (which contains elements of the "orignal" query we're building a plan for, and are thus against the API schema). When we later try to add the `a` field, it (somewhat reasonably) complains because that field is not in the API schema. I've pushed a fix at #2155 that ensure we build the selection against the subgraph schema properly in the first place. That PR also adds some more precise documentation on `udpateForAddingTo` that hopefully clarify the goals of that method and why it isn't the one having an issue here.
